R/SandwichLayerVariance.R

```diff
@@ -558,3 +559,459 @@ vcov_tee <- function(x, type = "CR0", cluster = NULL, ...) {
+#'  \code{teeMod} models with covariance adjustment and without absorbed effects
+#' @return a list of bread matrices
+#' @keywords internal
+.get_DB_covadj_bread <- function(x, ...) {
+  a11inv <- .get_a11_inverse(x)
+  a21 <- .get_a21(x)
+  a22inv <- .get_a22_inverse(x)
+  a21 <- matrix(0, nrow = nrow(a21), ncol = ncol(a21))
+  a22inv <- matrix(c(1,0,-1,1), nrow = 2, byrow = TRUE)
+  C <- matrix(c(1,0,0,1), nrow = 2, byrow = TRUE)
+  
+  n <- nrow(x$model)
+  bread1 <- a22inv %*% a21 %*% a11inv / n
+  bread2 <- - a22inv %*% C / n
+  
+  return(list(b1 = bread1, b2 = bread2))
+}
+
+#' @title (Internal) Meat matrix of design-based variance
+#' @param x a fitted \code{teeMod} model
+#' @details Calculate upper and lower bound estimates of meat matrix 
+#' for design-based variance estimate for \code{teeMod} models 
+#' with covariance adjustment and without absorbed effects
+#' @return a list of meat matrix bounds
+#' @keywords internal
+.get_DB_covadj_meat <- function(x, ...) {
+  agg <- .aggregate_to_cluster(x)
+  data <- agg[[1]]
+  bid <- data[, agg[[2]]]  # block ids
+  zobs <- data[, agg[[3]]] # observed zs
+  
+  p <- length(x$call$offset@fitted_covariance_model$coefficients)
+  XX <- .prepare_design_matrix(x)
+  
+  V00 <- .cov_mat_est(XX[zobs==0,], bid[zobs==0])
+  V11 <- .cov_mat_est(XX[zobs==1,], bid[zobs==1])
+  V01 <- .cov01_est(XX, zobs, bid)
+  
+  idl <- (p+2):(2*p+1)
+  meat1u <- V00[1:p, 1:p] + V01[1:p, 1:p] + t(V01[1:p, 1:p]) + V11[1:p, 1:p]
+  meat1l <- V00[idl, 1:p] + V01[idl, 1:p] + t(V01[idl, 1:p]) + V11[idl, 1:p]
+  
+  meat2u <- cbind(V00[1:p, p+1], V01[1:p, p+1]) + cbind(V01[p+1, 1:p], V11[1:p, p+1])
+  meat2l <- cbind(V00[idl, p+1], V01[idl, p+1]) + cbind(V01[2*p+2, 1:p], V11[idl, p+1])
+  
+  meat3u <- matrix(c(V00[p+1, p+1], V01[p+1, p+1], V01[p+1, p+1], V11[p+1, p+1]), ncol = 2)
+  meat3l <- matrix(c(V00[2*p+2, p+1], V01[2*p+2, p+1], V01[2*p+2, p+1], V11[2*p+2, p+1]), ncol = 2)
+  
+  return(list(m1u = meat1u, m1l = meat1l,
+              m2u = meat2u, m2l = meat2l,
+              m3u = meat3u, m3l = meat3l))
+}
+
+#' @title (Internal) Design-based variance for models without covariance adjustment
+#' @param x a fitted \code{teeMod} model
+#' @details Calculate bread matrix for design-based variance estimate for 
+#'  \code{teeMod} models without covariance adjustment and 
+#'  without absorbed effects
+#' @return design-based variance estimate of the main treatment effect estimate
+#' @keywords internal
+.get_DB_wo_covadj_se <- function(x, ...){
+  if (x@absorbed_intercepts) {
+    stop("x should not have absorbed intercepts")
+  }
+  if (!is.null(x$call$offset)){
+    stop("x should not have covariance adjustment")
+  }
+  agg <- .aggregate_to_cluster(x)
+  data <- agg[[1]]
+  block <- agg[[2]]
+  
+  ws <- data$.w
+  yobs <- data$.wy  # observed ys by cluster
+  zobs <- data[, agg[[3]]]  # observed zs by cluster
+  bid <- data[, block] # block ids of clusters
```

Comment:
Same note about using names of `agg`

Reply:
Updated in commit https://github.com/benbhansen-stats/propertee/commit/6eb8927664674088f61b3323683e7d592a2de431